File: demo-app/app/src/javaApp/java/com/launchkey/android/authenticator/demo/ui/activity/GenericFragmentDemoActivity.java
package com.launchkey.android.authenticator.demo.ui.activity;

import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.launchkey.android.authenticator.demo.R;
import com.launchkey.android.authenticator.demo.databinding.DemoActivityFragmentBinding;
import com.launchkey.android.authenticator.demo.ui.fragment.CustomLinkingFragment;
import com.launchkey.android.authenticator.sdk.core.authentication_management.Device;
import com.launchkey.android.authenticator.sdk.core.authentication_management.event_callback.UnlinkDeviceEventCallback;
import com.launchkey.android.authenticator.sdk.core.exception.DeviceUnlinkedButFailedToNotifyServerException;

public class GenericFragmentDemoActivity extends BaseDemoActivity<DemoActivityFragmentBinding> {

    public static final String EXTRA_TITLE = "title";
    public static final String EXTRA_FRAGMENT_CLASS = "fragment_class";
    public static final String EXTRA_GO_BACK_ON_UNLINK = "go_back_on_unlink";

    public GenericFragmentDemoActivity() {
        super(R.layout.demo_activity_fragment);
    }

    private boolean goBackOnUnlink;
    private final @NonNull UnlinkDeviceEventCallback unlinkDeviceEventCallback = new UnlinkDeviceEventCallback() {
        @Override
        public void onSuccess(final @NonNull Device device) {
            if (device.isCurrent() && goBackOnUnlink) {
                finish();
            }
        }

        @Override
        public void onFailure(final @NonNull Exception e) {
            if (e instanceof DeviceUnlinkedButFailedToNotifyServerException && goBackOnUnlink) {
                finish();
            }
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = DemoActivityFragmentBinding.bind(findViewById(R.id.demo_activity_fragment_root));

        setSupportActionBar(binding.demoActivityFragmentToolbar);

        Intent i = getIntent();
        Bundle extras = i == null ? null : i.getExtras();
        if (getSupportActionBar() != null) {
            String title = extras == null ? null : extras.getString(EXTRA_TITLE);
            getSupportActionBar().setTitle(title == null ? "Demo" : title);
        }

        if (savedInstanceState != null) {
            return;
        }

        String fragmentClass = extras == null ? null : extras.getString(EXTRA_FRAGMENT_CLASS);

        if (fragmentClass == null) {
            finish();
            return;
        }

        goBackOnUnlink = extras.getBoolean(EXTRA_GO_BACK_ON_UNLINK);

        //Instantiate the Fragment by name that was passed via extra (Bundle) and if not null,
        // then place it in the container.

        Fragment f;

        try {
            f = getSupportFragmentManager().getFragmentFactory().instantiate(getClassLoader(), fragmentClass);
            f.setArguments(extras);
        } catch (Exception e) {
            finish();
            return;
        }

        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.demo_activity_fragment_container, f)
                .commit();
    }

    @Override
    public void onResume() {
        super.onResume();
        getAuthenticatorManager().registerForEvents(unlinkDeviceEventCallback);
    }

    @Override
    public void onPause() {
        super.onPause();
        getAuthenticatorManager().unregisterForEvents(unlinkDeviceEventCallback);
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/activity/GenericFragmentDemoActivity.kt
package com.launchkey.android.authenticator.demo.ui.activity

import android.os.Bundle
import androidx.fragment.app.Fragment
import com.launchkey.android.authenticator.demo.R
import com.launchkey.android.authenticator.demo.databinding.DemoActivityFragmentBinding
import com.launchkey.android.authenticator.sdk.core.authentication_management.Device
import com.launchkey.android.authenticator.sdk.core.authentication_management.event_callback.UnlinkDeviceEventCallback
import com.launchkey.android.authenticator.sdk.core.exception.DeviceUnlinkedButFailedToNotifyServerException

class GenericFragmentDemoActivity : BaseDemoActivity<DemoActivityFragmentBinding>(R.layout.demo_activity_fragment) {
    private var goBackOnUnlink = false
    private val unlinkDeviceEventCallback: UnlinkDeviceEventCallback = object : UnlinkDeviceEventCallback() {
        override fun onSuccess(device: Device) {
            if (device.isCurrent && goBackOnUnlink) {
                finish()
            }
        }

        override fun onFailure(e: Exception) {
            if (e is DeviceUnlinkedButFailedToNotifyServerException && goBackOnUnlink) {
                finish()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DemoActivityFragmentBinding.bind(findViewById(R.id.demo_activity_fragment_root))
        setSupportActionBar(binding!!.demoActivityFragmentToolbar)
        val i = intent
        val extras = i?.extras
        if (supportActionBar != null) {
            val title = extras?.getString(EXTRA_TITLE)
            supportActionBar!!.title = title ?: "Demo"
        }
        if (savedInstanceState != null) {
            return
        }
        val fragmentClass = extras?.getString(EXTRA_FRAGMENT_CLASS)
        if (fragmentClass == null) {
            finish()
            return
        }
        goBackOnUnlink = extras.getBoolean(EXTRA_GO_BACK_ON_UNLINK)

        //Instantiate the Fragment by name that was passed via extra (Bundle) and if not null,
        // then place it in the container.
        val f: Fragment
        f = try {
            Fragment.instantiate(this, fragmentClass)
        } catch (e: Exception) {
            finish()
            return
        }
        supportFragmentManager
                .beginTransaction()
                .add(R.id.demo_activity_fragment_container, f)
                .commit()
    }

    public override fun onResume() {
        super.onResume()
        authenticatorManager.registerForEvents(unlinkDeviceEventCallback)
    }

    public override fun onPause() {
        super.onPause()
        authenticatorManager.unregisterForEvents(unlinkDeviceEventCallback)
    }

    companion object {
        const val EXTRA_TITLE = "title"
        const val EXTRA_FRAGMENT_CLASS = "fragment_class"
        const val EXTRA_GO_BACK_ON_UNLINK = "go_back_on_unlink"
    }
}